### PR TITLE
[ESM] Improve SQS batch collection and flushing

### PR DESCRIPTION
**Pull request by gregfurman**
## Motivation
This PR allows for the ESM poller to collect and send batches of SQS messages based on the number of messages collected and the duration of the batching event (up to 300s).

The motivations were as follows:
1. The feature request https://github.com/localstack/localstack/issues/5042 required LocalStack to support batch size over 10, which in turn would depend on [MaximumBatchingWindowInSeconds](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html#cfn-lambda-eventsourcemapping-maximumbatchingwindowinseconds) being larger than 1. While offering CRUD support for this parameter, LocalStack did not actually collect a batch for the specified duration. Hence, these functional changes mean that we can now collect records until the window elapses.

2. Under high volumes of requests, we expect the latency of LocalStack's gateway to degrade due. By supporting window batching, we have also introduced the ability for ESM pollers to long-poll an SQS event source for batches of data -- which should help alleviate some pressure on the gateway. 

## Changes
- Adds long-polling for SQS ReceiveMessage that will continuously poll an SQS queue until either a desired number of messages BatchSize has been collected or the minimum between MaximumBatchingWindowInSeconds has elapsed.
- Once collected, this batch is potentially further split into chunks of up to 2.5K records each in lieu of proper flushing based on Lambda payload quotas (see [docs](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventsourcemapping.html#invocation-eventsourcemapping-batching)).

## Testing
- Unskips the `test_sqs_event_source_mapping_batching_reserved_concurrency` since we can now collect and send batches of more than 10 records at a time.

## Benchmarking

Some benchmarks were run to assess how the long-polling changes would the latency of the LocalStack gateway. See benchmarking results [here](https://github.com/localstack/locust-benchmarking/tree/experiment/long-polling/experiments/long_polling_esm_sqs).

tl;dr Long polling saw fewer requests being made and, in some cases, had improved latencies at higher percentiles `P(>95)`.

### Assumptions

- A higher load on the gateway means fewer available threads for processing new requests. Therefore, we'd expect short-polling calls with many poll-misses to add unnecessary load to the gateway (due to each processing call requiring a worker) that equivalent long-polling calls should be able to circumvent.
- We poll an event source at an interval of once per second. Each long-polling call will block a single gateway thread (there are 1K available) while waiting for a request to complete. Therefore, we expect that (in extremis) many ESM pollers all running long-polling calls will cause resource contention and degrade performance.
- For most cases, however, we assume that long-polling calls should reduce the number of requests made to the gateway within a given duration, therefore improving the availability of workers to process _other_ requests.

### Results

- Performance with and without long polling was fairly similar in terms of latency across all experiments, with differences of between 50-300ms between the two approaches.
- Large batch sizes coupled with high batch windows saw the largest improvements in performance.
- Long polling resulted in fewer requests (`~7%`) being processed within the 5m test interval - likely due to high resource contention while long-polling calls were being made. However, this did not hold true for maximising Batch Size (`10k`) and Batch Window (`300s`) with long-polling which saw the highest throughput and best latency improvements.

### Conclusions

- The effects of long-polling were sometimes seen at higher latency percentiles -- where long polling runs had less latency than short polling. However, no notable changes in performance were observed between long and short polling.
- Performance was optimized in both cases by using larger batch sizes and longer batch windows, with the combination of long polling + large batches + long windows showing the best overall efficiency.
- Importantly, while introducing long-polling did not dramatically effect performance (positively or negatively), this provides LocalStack with an avenue for optimising

### CODE REVIEW

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 1]
```diff
@@ -57,22 +62,88 @@ def get_queue_attributes(self) -> dict:
...
+        messages = self.collect_messages(
+            max_batch_size=self.sqs_queue_parameters["BatchSize"],
+            max_batch_window=self.sqs_queue_parameters["MaximumBatchingWindowInSeconds"],
         )
-        if messages := response.get("Messages"):
+
+        # NOTE: If a batch is collected, this will send a single collected batch for each poll call.
+        # Increasing the poller frequency _should_ influence the rate of collection but this has not
+        # yet been investigated.
+        # messages = next(self.collector)
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:145]: Needs to be removed... will wait for CI to pass before doing this.

```suggestion
```

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 2]
```diff
@@ -57,22 +62,87 @@ def get_queue_attributes(self) -> dict:
...
+
+            # Return the messages received and the request duration in seconds.
+            try:
+                messages, request_duration = receive_message(messages_per_receive)
+            except Exception as e:
+                # If an exception is raised here, break the loop and return whatever
+                # has been collected early.
+                # TODO: Handle exceptions differently i.e QueueNotExist or ConnectionFailed should retry with backoff
+                LOG.warning(
+                    "Polling SQS queue failed: %s",
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:106]: nit: Can we add the queue name (`self.source_arn`) in the error message to simplify debugging

**Review by dfangl — COMMENTED** [review]
This already looks quite good, but i still have a concern:

We do not have an upper limit for the sleep times. As per your table, a queue which gets 1 message/s, and a batching window of 300s set, and a batch size of 1, triggers only once every 150s (on average, with no latency), which causes a massive queue buildup. If someone sets a 300s batch window, and a burst of messages occurs, we should still process them timely. I think AWS would do so as well.

Also please add unit descriptions to the table, as it makes it clearer :)

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 3]
```diff
@@ -57,22 +62,65 @@ def get_queue_attributes(self) -> dict:
...
 
+    def collect_messages(self, max_batch_size=10, max_batch_window=0, **kwargs) -> list[dict]:
+        # TODO: Set to max_batch_size when override message count changes are merged.
+        messages_per_receive = min(DEFAULT_MAX_RECEIVE_COUNT, max_batch_size)
+
+        # Number of messages we want to receive per ReceiveMessage operation.
+        def receive_message(num_messages: int = messages_per_receive):
+            response = self.source_client.receive_message(
+                QueueUrl=self.queue_url,
+                MaxNumberOfMessages=num_messages,
```
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:73]: This `receive_message` call is going to leverage the beyond 10 limits implemented via the header-based approach: https://github.com/localstack/localstack/pull/12118
We should make this more visible, at least through some documentation/comments.

**Review by joe4dev — COMMENTED** [review]
The simplified message collection is much clearer 👏  Thank you for these refinements @gregfurman 

I'm looking forward to supporting batching windows and batching by size 🙌 
However, I'm concerned about flooding the LS gateway within the while loop (see ⚠️). How do you assess the scenario outlined in the comment?

**Review by joe4dev — COMMENTED** [review]
Thank you for your relentless efforts toward solving this tricky batching challenge @gregfurman 🙇 

As discussed with @gregfurman and @bentsku, long polling has several advantages (e.g., responsiveness, no gateway flooding, AWS behavior parity). However, I don't fully understand the limitations (e.g., are blocked gateway threads acceptable?, does LS shutdown gracefully?) and would love to get @dfangl opinion on this one, especially given the potentially big blast radius 💯 and experience (didn't we switch from long to short polling at some point 🤔 ).

File: localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py [hunk 4]
```diff
@@ -111,6 +114,17 @@ def get_esm_worker(self) -> EsmWorker:
                 role_arn=self.function_role_arn,
                 service_principal=ServicePrincipal.lambda_,
                 source_arn=self.esm_config["FunctionArn"],
+                client_config=botocore.config.Config(
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py:117]: Very true. Will open a PR. I'm going to set this to `300s` in pipes.

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 5]
```diff
@@ -97,28 +108,47 @@ def event_source(self) -> str:
...
+                "Polling SQS queue %s failed: %s",
+                self.source_arn,
+                e,
+                exc_info=LOG.isEnabledFor(logging.DEBUG),
+            )
+        messages = response.get("Messages", [])
+        # NOTE: If the collection of messages exceeds the 6MB size-limit imposed on payloads sent to a Lambda,
+        # split into chunks of up to 6MB each.
+        # See https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventsourcemapping.html#invocation-eventsourcemapping-batching
+        for message_batch in batched_by_size(messages, 5e6):
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:142]: > Shouldn't this fix this test skip? 

That `batched_by_size` does not account for the size of the JSON object the dict will be marshalled into.

I've added a limit of 2.5K records to be sent to a target Lambda at a time since I think this 6MB limit actually falls out-of-scope. The test skip can be reverted now 😄 

> Do we have a test for size-based batch flushing?

We have a test in `tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py::TestSQSEventSourceMapping::test_sqs_event_source_mapping_batch_size_override`

File: localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py [hunk 6]
```diff
@@ -111,6 +114,17 @@ def get_esm_worker(self) -> EsmWorker:
...
+                    retries={"total_max_attempts": 1},  # Disable retries
+                    read_timeout=max(
+                        self.esm_config.get(
+                            "MaximumBatchingWindowInSeconds", DEFAULT_MAX_WAIT_TIME_SECONDS
+                        ),
+                        60,
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py:123]: > This configuration caps the read_timeout at 60 seconds. Aren't we running into read timeouts whenever the MaximumBatchingWindowInSeconds exceeds 60s (+5 buffer)?

This configuration sets the `read_timeout` to the larger value between `MaximumBatchingWindowInSeconds` and `60s` -- so we shouldn't encounter issues with timeouts.

> The [MaximumBatchingWindowInSeconds](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html#cfn-lambda-eventsourcemapping-maximumbatchingwindowinseconds) can be up to 300s (5min), which is used in the header-based wait time override here: 

I've created a seperate PR for this now -- will remove the functionality from the SQS provider. See https://github.com/localstack/localstack/pull/12172

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 7]
```diff
@@ -54,13 +61,17 @@ def _register_client_hooks(self):
...
+            if requested_count and requested_count > DEFAULT_MAX_RECEIVE_COUNT:
+                # Allow overide parameter to be greater than default and less than maximum batch size.
+                # Useful for getting remaining records less than the batch size. i.e we need 100 records but BatchSize is 1k.
+                override = min(requested_count, self.sqs_queue_parameters["BatchSize"])
+                context[HEADER_LOCALSTACK_SQS_OVERRIDE_MESSAGE_COUNT] = str(override)
+                params["MaxNumberOfMessages"] = DEFAULT_MAX_RECEIVE_COUNT
+
+            requested_wait_time = params.get("MaximumBatchingWindowInSeconds")
+            if requested_wait_time and requested_wait_time > DEFAULT_MAX_WAIT_TIME_SECONDS:
+                context[HEADER_LOCALSTACK_SQS_OVERRIDE_WAIT_TIME_SECONDS] = str(requested_wait_time)
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:73]: Yeah exactly!

File: localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py [hunk 8]
```diff
@@ -111,6 +114,17 @@ def get_esm_worker(self) -> EsmWorker:
...
+                    )
+                    + 5,  # Extend read timeout (with 5s buffer) for long-polling
+                    tcp_keepalive=True,
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py:126]: doc: some motivation and reference would be helpful on why we're doing this
Nice AWS reference: https://aws.amazon.com/blogs/networking-and-content-delivery/implementing-long-running-tcp-connections-within-vpc-networking/

File: localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (joe4dev) [reply localstack-core/localstack/services/lambda_/event_source_mapping/esm_worker_factory.py:117]: Done in https://github.com/localstack/localstack-ext/pull/3970

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 9]
```diff
@@ -100,28 +129,45 @@ def event_source(self) -> str:
...
             # Override how many messages we can receive per call
             sqs_override_max_message_count=self.batch_size,
+            # Override how long to wait until batching conditions are met
+            sqs_override_wait_time_seconds=self.maximum_batching_window,
         )
-        if messages := response.get("Messages"):
-            LOG.debug("Polled %d events from %s", len(messages), self.source_arn)
+
+        messages = response.get("Messages", [])
+        LOG.debug("Polled %d events from %s", len(messages), self.source_arn)
```
[the remark this answers is not in the document]
Comment (gregfurman) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:151]: I was thinking it could be useful for debugging to log explicitly whether nothing was polled from the event source. Perhaps we can distinguish this better with a `"Polled no events from %s"` -- wdyt?

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 10]
```diff
@@ -98,30 +127,67 @@ def event_source(self) -> str:
...
+        # * Higher batch collection durations and no. of records retrieved per request mean fewer calls to the LocalStack gateway [3] when polling an event-source [4].
+        # * LocalStack shutdown works because the LocalStack gateway shuts down and terminates the open connection.
+        # * Provider lifecycle hooks have been added to ensure blocking long-poll calls are gracefully interrupted and returned.
+        #
+        # Benchmarking showed the long-polling optimizations (with a 20s duration) improved LS RPS by >100%:
+        # * Short-polling: 282.26 req/sec
+        # * Long-polling: 576.32 req/sec (~2.04x improvement)
+        #
+        # Pros (+) / Cons (-):
+        # + Reduces latency because the `ReceiveMessage` call immediately returns once we reach the desired `BatchSize` or the `WaitTimeSeconds` elapses.
```
Comment (gregfurman) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:141]:
```suggestion
        # + Alleviates pressure on the gateway since each `ReceiveMessage` call only returns once we reach the desired `BatchSize` or the `WaitTimeSeconds` elapses.
```

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 11]
```diff
@@ -133,12 +133,8 @@ def poll_events(self) -> None:
         # * LocalStack shutdown works because the LocalStack gateway shuts down and terminates the open connection.
         # * Provider lifecycle hooks have been added to ensure blocking long-poll calls are gracefully interrupted and returned.
         #
-        # Benchmarking showed the long-polling optimizations (with a 20s duration) improved LS RPS by >100%:
-        # * Short-polling: 282.26 req/sec
-        # * Long-polling: 576.32 req/sec (~2.04x improvement)
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py]: Thank you for removing these misleading results 👍

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py [hunk 12]
```diff
@@ -133,12 +133,8 @@ def poll_events(self) -> None:
...
-        #
         # Pros (+) / Cons (-):
-        # + Reduces latency because the `ReceiveMessage` call immediately returns once we reach the desired `BatchSize` or the `WaitTimeSeconds` elapses.
+        # + Alleviates pressure on the gateway since each `ReceiveMessage` call only returns once we reach the desired `BatchSize` or the `WaitTimeSeconds` elapses.
```
Comment (joe4dev) [inline localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:137]: Good clarification ✨

**Review by joe4dev — APPROVED** [review]
Thank you for the unlimited persistence and going the extra mile @gregfurman to gain more confidence and learn more about potential limitations of this impactful change 🙇 

The code changes look good, we have iterated on them and all suggestions are taken into consideration 👏👏👏 .

The benchmark results indicate that this change does not negatively impact the LocalStack gateway latency, but we should keep an 👂 open to potential user feedback in large (>100 ESMs) and high-performance (>100 users) environments.

I love the out-of-the-box thinking around custom headers to go beyond AWS API restrictions to reduce the number of requests on LocalStack while still keeping remote API compatibility 🧠 💯 

Strictly speaking, we don't have explicit aws-validated tests for the new `MaximumBatchingWindowInSeconds` feature, but I think that's fine here given how the cost/value trade-off. It would be hard to test against AWS because their internal behavior is somewhat unpredictable (depending on internal performance optimizations) and probably not guaranteed (i.e., max is an upper bound). It would likely require multiple iterations of time-consuming testing.


Good to 🚢  from my side 🚀 


nit: Any idea where that weird line mismatch (`995` vs `999`) comes from in the [failing Lambda test](https://app.circleci.com/pipelines/github/localstack/localstack/31269/workflows/0ff8a1eb-9513-4029-b080-efe95b0e9ded/jobs/278699/tests)?
It's clearly unrelated to these changes.

File: localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dfangl) [reply localstack-core/localstack/services/lambda_/event_source_mapping/pollers/sqs_poller.py:166]: I think we should be very careful about this - many people have `DEBUG=1` by default, and this can be a lot. I agree it is not urgent to remove - but especially the message for no events could be removed in the future.

File: tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py [hunk 13]
```diff
@@ -1617,17 +1608,72 @@ def test_sqs_event_source_mapping_batch_size_override(
         cleanups.append(lambda: aws_client.lambda_.delete_event_source_mapping(UUID=mapping_uuid))
         _await_event_source_mapping_enabled(aws_client.lambda_, mapping_uuid)
 
+        expected_invocations = -(batch_size // -2500)  # converts floor division to ceil
```
Comment (dfangl) [inline tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py:1611]: Couldn't we just add `math.ceil` here? If I have to explain what a line does in a comment, it is not really well readable :)

**Review by dfangl — APPROVED** [review]
I think this is good to go, let's get it in!
I agree with Joel's final commend wholeheartedly, great work and patience with this PR!

File: tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py (hunk 13 above)
Comment (gregfurman) [reply tests/aws/services/lambda_/event_source_mapping/test_lambda_integration_sqs.py:1611]: Lol yeah. Lemme change this now...

**Comment by joe4dev** [comment]
@gregfurman Do we know why `test_sqs_event_source_mapping_batching_reserved_concurrency` fails? Is it flaky?
https://app.circleci.com/pipelines/github/localstack/localstack/31294/workflows/85e2e6dd-0f6e-4014-9612-2bbad5b4e266/jobs/278937/tests
